### ISSUE TITLE
Migrate configuration system to Hydra + OmegaConf

### ISSUE THREAD

**Issue by aymuos15** [aymuos15/nnBenchmark#30]
## Summary

Replace the hand-rolled argparse + PyYAML + custom merge/validation config system with [Hydra](https://hydra.cc/) and OmegaConf. This gives us typed structured configs, CLI overrides for free, config composition (model/loss/optimizer groups), and multi-run sweeps — critical for a benchmarking framework.

Also replaces the 500-line `yaml_generator.py` (manual `f.write()` calls) with structured config serialization.

## Motivation

**Current pain points:**

1. **`yaml_generator.py`** is 500+ lines of `f.write()` string formatting — fragile, hard to maintain, easy to introduce YAML syntax bugs
2. **Custom config inheritance** (`config/merge.py`) reimplements what OmegaConf does natively with `OmegaConf.merge()`
3. **Custom config validation** (`config/validation.py`) manually checks field existence — structured configs do this automatically
4. **Custom path resolution** (`config/resolution.py`) manually resolves config paths — Hydra has native config search paths
5. **No CLI overrides** — users must edit YAML files to change a single parameter; Hydra gives `nnBench.train training.epochs=500` for free
6. **No sweep support** — benchmarking needs `nnBench.train -m dataset.fold=0,1,2,3,4` which Hydra provides out of the box

**Current architecture:**
```
CLI (argparse) → yaml.safe_load → dict[str, Any] → cfg["key"]["subkey"] everywhere
```

**Target architecture:**
```
CLI (Hydra @hydra.main) → OmegaConf DictConfig → structured config dataclasses → typed access
```

### Config system (`src/config/`)
| File | Action | Notes |
|---|---|---|
| `config/load.py` | **Rewrite** | `yaml.safe_load` → `OmegaConf.load()` / Hydra loading |
| `config/merge.py` | **Delete** | `deep_merge_dicts` / `validate_override_keys` → `OmegaConf.merge()` |
| `config/resolution.py` | **Delete** | Path resolution → Hydra config search paths |
| `config/validation.py` | **Delete** | Field validation → structured config `__post_init__` methods |
| `config/paths.py` | **Keep** | Env var resolution (`nnBench_raw`, etc.) — register as OmegaConf resolvers |
| `config/__init__.py` | **Update** | Update re-exports |

### Config generation (`src/planning/`)
| File | Action | Notes |
|---|---|---|
| `planning/yaml_generator.py` | **Rewrite** | 500 lines of `f.write()` → build structured config + `OmegaConf.to_yaml()` |

### CLIs (5 entry points)
| File | Action | Notes |
|---|---|---|
| `planning/cli.py` | **Rewrite** | argparse → `@hydra.main` |
| `engines/train/cli.py` | **Rewrite** | argparse → `@hydra.main` |
| `engines/validate/cli.py` | **Rewrite** | argparse → `@hydra.main` |
| `engines/inference/cli.py` | **Rewrite** | argparse → `@hydra.main` |
| `plotting/cli.py` | **Rewrite** | argparse → `@hydra.main` |

### Config consumers (~60 sites in `src/engines/`)
| File | Action | Notes |
|---|---|---|
| `engines/setup.py` | **Update** | `cfg["dataset"]["name"]` access patterns |
| `engines/train/run.py` | **Update** | ~15 `cfg["key"]` access sites |
| `engines/validate/run.py` | **Update** | ~10 `cfg["key"]` access sites |
| `engines/inference/run.py` | **Update** | ~8 `cfg["key"]` access sites |
| `engines/inference/engine.py` | **Update** | `cfg["dataset"]["num_classes"]` etc. |
| `engines/ignite_utils/trainer.py` | **Update** | `cfg["model"]`, `cfg["training"]`, `cfg["optimizer"]`, `cfg["loss"]` |
| `engines/validate/handlers.py` | **Update** | `cfg["dataset"]["name"]` |
| `engines/inference/handlers.py` | **Update** | `cfg["dataset"]["name"]` |

### Tests
| File | Action | Notes |
|---|---|---|
| `tests/test_planning.py` | **Update** | `TestYAMLGenerator` class, `yaml.safe_load` assertions |
| `tests/test_config.py` | **Update** | Config loading tests |
| `tests/test_config_merge.py` | **Delete or rewrite** | Tests custom merge logic that gets removed |
| `tests/test_config_validation.py` | **Update** | Validation tests → structured config tests |
| `tests/test_engines_train_run.py` | **Update** | Creates configs via `yaml.dump` (still works) |
| `tests/test_engines_validate_run.py` | **Update** | Creates configs via `yaml.dump` (still works) |
| `tests/conftest.py` | **Update** | `sample_config` fixture creates configs via `yaml.dump` |

### Phase 0: Add dependency (no behavior change)
- Add `hydra-core>=1.3` and `omegaconf>=2.3` to `pyproject.toml` dependencies
- Verify clean install alongside existing deps (torch, monai, etc.)

### Phase 1: Structured config dataclasses (additive, no breakage)
Create `src/config/schema.py` with nested frozen dataclasses mirroring the YAML structure:

```python
from dataclasses import dataclass, field
from omegaconf import MISSING

@dataclass
class CacheConfig:
    enabled: bool = False
    cache_rate: float = 0.0

@dataclass
class DatasetConfig:
    name: str = MISSING
    median_shape: list[int] = MISSING
    median_spacing: list[float] = MISSING
    foreground_intensity_mean: float = MISSING
    spatial_size: list[int] = MISSING
    num_classes: int = MISSING
    fold: int = 0
    cache: CacheConfig = field(default_factory=CacheConfig)

@dataclass
class DynUNetConfig:
    filters: list[int] = MISSING
    kernel_size: list[list[int]] = MISSING
    strides: list[list[int]] = MISSING
    upsample_kernel_size: list[list[int]] = MISSING
    norm_name: Any = field(default_factory=lambda: ["INSTANCE", {"affine": True}])
    act_name: Any = field(default_factory=lambda: ["leakyrelu", {"inplace": True, "negative_slope": 0.01}])
    res_block: bool = False

@dataclass
class UNetConfig:
    channels: list[int] = MISSING
    strides: list[int] = MISSING
    num_res_units: int = 2

@dataclass
class ModelConfig:
    type: str = "NativeDSDynUNet"
    spatial_dims: int = 3
    in_channels: int = MISSING
    out_channels: int = MISSING
    deep_supervision: bool = True
    deep_supr_num: int = MISSING
    ds_weights: list[float] = MISSING
    DynUNet: DynUNetConfig = field(default_factory=DynUNetConfig)
    UNet: UNetConfig = field(default_factory=UNetConfig)

@dataclass
class TrainingConfig:
    epochs: int = 1000
    num_iterations_per_epoch: int = 250
    batch_size: int = MISSING
    learning_rate: float = 0.01
    val_interval: int = 1
    num_workers: int = 0
    checkpoint_metric: str = "DiceMetric"
    mixed_precision: bool = True

@dataclass
class OptimizerConfig:
    type: str = "SGD"
    weight_decay: float = 0.00003
    momentum: float = 0.99
    nesterov: bool = True

@dataclass
class LossConfig:
    type: str = "DiceCELoss"
    to_onehot_y: bool = True
    softmax: bool = True
    batch: bool = False

@dataclass
class NnBenchConfig:
    _comment: str = ""
    dataset: DatasetConfig = field(default_factory=DatasetConfig)
    model: ModelConfig = field(default_factory=ModelConfig)
    training: TrainingConfig = field(default_factory=TrainingConfig)
    optimizer: OptimizerConfig = field(default_factory=OptimizerConfig)
    loss: LossConfig = field(default_factory=LossConfig)
    validation_metrics: list[Any] = MISSING
    inference_metrics: list[Any] = MISSING
    transforms: Any = MISSING
    inference: Any = field(default_factory=lambda: {"batch_size": 1})
```

This phase is **pure addition** — no existing code changes. Dataclasses serve as documentation and are independently testable.

### Phase 2: Replace `yaml_generator.py`
The generator becomes:
1. Build `NnBenchConfig` from `ExperimentPlan` (structured, type-checked)
2. `OmegaConf.to_yaml(cfg)` → write to file

500 lines → ~80 lines. Inline YAML comments are lost; if those matter for new users, use a Jinja2 template that renders from the structured config (data assembly is still type-safe).

### Phase 3: Replace config loading
- `load_config()` → `OmegaConf.load()` + `OmegaConf.merge()` for base_config inheritance
- Register env var resolvers for `nnBench_raw`, `nnBench_preprocessed`, `nnBench_results`:
  ```python
  OmegaConf.register_new_resolver("nnbench_raw", lambda: os.environ["nnBench_raw"])
  ```
- Delete `config/merge.py` — `OmegaConf.merge()` handles deep merge natively
- Move `validate_deep_supervision_config` / `validate_sliding_window_config` to `__post_init__` on structured config dataclasses
- Delete `config/validation.py`

### Phase 4: Migrate CLIs (one at a time)

**Migration order:** `train` → `validate` → `inference` → `plot` → `plan`

`train` goes first to validate the approach end-to-end. `plan` goes last because it generates configs rather than consuming them.

**Before (`engines/train/cli.py`):**
```python
def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--config", required=True)
    parser.add_argument("--dataset", default=None)
    parser.add_argument("--fresh", action="store_true")
    args = parser.parse_args()
    run_training(args.config, dataset=args.dataset, force_fresh=args.force_fresh)
```

**After:**
```python
@hydra.main(version_base=None, config_path="conf", config_name="config")
def main(cfg: NnBenchConfig) -> None:
    run_training(cfg)
```

**Key decisions:**
- **Config search path**: `conf/` directory at project root or use `--config-path` / `--config-name` CLI args
- **`nnBench_*` env vars**: Registered as OmegaConf resolvers — keeps existing env var convention
- **`--dataset` argument**: Becomes `dataset.name=Dataset001_Hippo` override
- **`--fresh` flag**: Becomes `+force_fresh=true` override or kept as a Hydra override group
- **Config resolution** (`config/resolution.py`): Hydra handles search paths natively — delete this module
- **Working directory**: Use `hydra.run.dir=.` to prevent Hydra from changing cwd (otherwise conflicts with `nnBench_results` convention)

### Phase 5: Update `cfg["key"]` access patterns (~60 sites)

`OmegaConf.DictConfig` supports **both** `cfg["key"]` and `cfg.key` access — so this phase is **optional for correctness** but improves developer experience.

**Recommendation:** Keep `cfg["key"]` working during migration (zero breakage), then incrementally migrate to `cfg.key` dot-access for IDE autocomplete and type checking.

### Phase 6: Config groups (future, post-migration)
Once Hydra is in place, add config groups:
```
conf/
├── config.yaml          # defaults list
├── model/
│   ├── dynunet.yaml
│   └── unet.yaml
├── loss/
│   ├── dice_ce.yaml
│   └── focal.yaml
└── optimizer/
    ├── sgd.yaml
    └── adam.yaml
```

Enables: `nnBench.train model=unet loss=focal`
Sweeps: `nnBench.train -m dataset.fold=0,1,2,3,4`

### Phase 7: Cleanup
Delete:
- [x] `src/config/merge.py`
- [x] `src/config/resolution.py`
- [x] `src/config/validation.py`
- [x] `tests/test_config_merge.py`
- [x] Argparse code from all 5 CLI modules

## Risks and Mitigations

| Risk | Impact | Mitigation |
|---|---|---|
| Hydra manages working directory | Breaks `nnBench_results` path convention | Set `hydra.run.dir=.` in default config |
| `OmegaConf.DictConfig` vs `dict` | `.copy()` returns DictConfig not dict | `OmegaConf.to_container(cfg)` where plain dict is needed (e.g., `model_cfg.pop("type")` patterns in `setup.py`) |
| Tests create configs via `yaml.dump` | Tests might break | `OmegaConf.load()` reads standard YAML — tests should still work |
| `pyproject.toml` `[project.scripts]` entry points | Hydra needs `@hydra.main` wrapper | Verified compatible — Hydra works with setuptools entry points |
| Comments lost in generated configs | Less user-friendly generated YAML | Use Jinja2 template for generation if comments are needed |
| `norm_name: [INSTANCE, {affine: true}]` mixed types | OmegaConf may struggle with `list[str | dict]` | Use `Any` type annotation for these fields |

## PR Sequence

| PR | Phases | Scope | Risk |
|---|---|---|---|
| **PR 1** | 0 + 1 | Add hydra dep + structured config dataclasses | None (additive only) |
| **PR 2** | 2 | Rewrite `yaml_generator.py` | Low (output format changes, loaded values identical) |
| **PR 3** | 3 + 4 (train only) | Replace config loading + migrate train CLI | Medium (validates approach end-to-end) |
| **PR 4** | 4 (validate + inference) | Migrate validate and inference CLIs | Low (pattern established) |
| **PR 5** | 4 (plot + plan) | Migrate remaining CLIs | Low |
| **PR 6** | 5 + 7 | Cleanup dict access + delete dead code | Low |
| **PR 7** | 6 | Add config groups + sweep support | None (additive) |

**Comment by aymuos15** [comment]
Closing in favor of a simpler MONAI Bundle-based migration. Hydra+OmegaConf is over-engineered for this project — we already use MONAI heavily and the safe_getattr pattern is a manual reimplementation of what ConfigParser does natively.
